### PR TITLE
Fix and extend counsel-file-jump

### LINKED ISSUE #1820: `counsel-jump-file`: Don't ignore '.*' files if search pattern begins with `\.`
Just wanted to see all my `.gitignore` files with `counsel-jump-file` but there were no results.  Since I'm using ``"\\(?:\\`[#.]\\)\\|\\(?:[#~]\\'\\)"`` as `counsel-find-file-ignore-regexp`, I quickly found this

https://github.com/abo-abo/swiper/blob/9414f7ab9b479cc6a66539b8f1494c81734d9acf/counsel.el#L1799

I believe this has to be extended for `\.` too as this is what you'd type for `counsel-jump-file` compared to `counsel-find-file`.

Furthermore, the following implicit filtering smells as well:

https://github.com/abo-abo/swiper/blob/9414f7ab9b479cc6a66539b8f1494c81734d9acf/counsel.el#L2373

I guess this will also shut down `.gitignore` files for example which is sad.  I'd let users decide what to filter.

### PR DESCRIPTION
**Pull request by mookid**
+ avoid relying on shell expansion
+ port to windows via dir.exe
+ make the subdirectories blacklist customizable

fix #1820

### CODE REVIEW

File: counsel.el [hunk 1]
```diff
@@ -2426,6 +2426,26 @@ FZF-PROMPT, if non-nil, is passed as `ivy-read' prompt argument."
...
+        ((memq system-type '(ms-dos windows-nt))
+         (mapcar (lambda (fullpath)
+                   (string-remove-prefix dir
+                                         (expand-file-name fullpath)))
+                 (split-string
+                  (shell-command-to-string
+                   "dir /s/b/A:-D | findstr /V .git\\")
+                  "\n" t)))
+        (t
+         (error "counsel--file-jump-function: not compatible"))))
```
[the remark this answers is not in the document]
Comment (basil-conto) [reply counsel.el]: The author is @abo-abo, I just had some further suggestions. FWIW I'll try to catch up with everything after my holiday visiting family.

[the remark this answers is not in the document]
Comment (mookid) [reply counsel.el]: @Alexander-Shukaev I don't think that find (or directory) is slow enough to justify the additional complexity. 

My point really is to make it work everywhere in a uniform way for everyone (esp. Windows users).

I agree that generalising to VCs other than git is an interesting extension, but I won't add support for the currently trending tool, at least in this PR.

**Comment by mookid** [comment]
@Alexander-Shukaev see the last commit for your feature request.

**Comment by Alexander-Shukaev** [comment]
I really don't want to sound pedantic but the name `counsel-file-jump-ignored-directories` does not look well when there is also `dired-jump` where it is used.  Just take a general name `counsel-find-ignored-directories` or `counsel-ignored-directories`; chances are this variable will come handy in other functions as well.  Handwritten default value also seems unnecessary and opinionated.  Why not just use `vc-directory-exclusion-list` or directly `grep-find-ignored-directories`?  They are both richer, standardized, and maintained by someone else (who perhaps also knows better what the defaults should be).

File: counsel.el [hunk 2]
```diff
@@ -2425,6 +2425,73 @@ FZF-PROMPT, if non-nil, is passed as `ivy-read' prompt argument."
...
+
+(defun counsel--dir.exe-dir-mask (dirs)
+  (concat "/V "
+          (mapconcat
+           (lambda (d) (concat "/C:" d "\\"))
+           dirs
+           " ")))
+
+(defun counsel--list-files (find-command dir-command dir)
+  (cond ((executable-find find-program)
```
Comment (mookid) [inline counsel.el]: TODO: on windows, this succeeds and find another `find` that is not useful.

**Comment by Alexander-Shukaev** [comment]
Why is this closed?

**Pull request closed without merging**